src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,253 @@
+            return numberOfBytes;
+        }
+
+        public static int NumberOfWrittenBytes(InteropBio.BioHandle handle)
+        {
+            var bio = (bio_st*)handle.Handle;
+            var bytes = (int)bio[0].num_write;
+            bio[0].num_write = 0;
+            return bytes;
+        }
+
+        private static int ReadBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var buffer = Unsafe.Read<ReadableBuffer>(bio.ptr);
+            if (buffer.Length == 0)
+            {
+                bio.num = 0;
+                if (numberOfBytes > 0)
+                {
+                    BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_READ | BioFlags.BIO_FLAGS_SHOULD_RETRY);
+                }
+                return -1;
+            }
+            if (numberOfBytes == 0)
+            {
+                return 0;
+            }
+            numberOfBytes = Math.Min(numberOfBytes, buffer.Length);
+            buffer.Slice(0, numberOfBytes).CopyTo(new Span<byte>(buff, numberOfBytes));
+            Unsafe.Write(bio.ptr, buffer.Slice(numberOfBytes));
+            bio.num = buffer.Length - numberOfBytes;
+            return numberOfBytes;
+        }
+
+        public static void SetReadBufferPointer(InteropBio.BioHandle bio, ref ReadableBuffer buffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].num = buffer.Length;
+            b[0].ptr = Unsafe.AsPointer(ref buffer);
+        }
+
+        public static void SetWriteBufferPointer(InteropBio.BioHandle bio, GCHandle handleToWriteBuffer)
+        {
+            var b = (bio_st*)bio.Handle;
+            b[0].ptr = (void*)GCHandle.ToIntPtr(handleToWriteBuffer);
+        }
+
+        private static long ControlBio(ref bio_st bio, BioControl cmd, long num, void* ptr)
+        {
+            switch (cmd)
+            {
+                case BioControl.BIO_CTRL_FLUSH:
+                case BioControl.BIO_CTRL_POP:
+                case BioControl.BIO_CTRL_PUSH:
+                    return 1;
+            }
+            return 0;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct bio_method_st
+        {
+            public int type;
+            public void* name;
+            public Write bwriteDelegate; // int (*bwrite) (BIO*, const char*, int);
+            public Read breadDelegate; //int (*bread) (BIO*, char*, int);
+            public void* bputsDelegate; //int (*bputs) (BIO*, const char*);
+            public void* bgestsDelegate; //int (*bgets) (BIO*, char*, int);
+            public Control ctrlDelegate; //long (*ctrl) (BIO*, int, long, void*);
+            public Create create; //int (*create) (BIO*);
+            public Free destroy; //int (*destroy) (BIO*);
+            public void* callback_ctrl; //long (*callback_ctrl) (BIO*, int, bio_info_cb*);
+        }
+        [StructLayout(LayoutKind.Sequential)]
+        struct bio_st
+        {
+            public void* method;
```

Comment:
The delegates that are impemented are strongly typed, the rest are not applicable as the BIO definition covers files, sockets and all sorts of other IO. So in answer the ones used are auto marshaled and strongly typed.